File: python/cache_manager.py
# python/cache_manager.py

import pickle
import os
from typing import Any, Optional
from python.logger import logger  # Import the centralized logger

def load_cache(cache_file_path: str) -> Optional[Any]:
    """
    Loads data from a cache file if it exists.

    Args:
        cache_file_path (str): Path to the cache file.

    Returns:
        Optional[Any]: Loaded data or None if loading fails.
    """
    if os.path.exists(cache_file_path):
        logger.info(f"Loading cached data from {cache_file_path}.")
        try:
            with open(cache_file_path, 'rb') as cache_file:
                data = pickle.load(cache_file)
            logger.info(f"Cached data loaded successfully from {cache_file_path}.")
            return data
        except Exception as e:
            logger.error(f"Error loading cache from {cache_file_path}: {e}")
            return None
    else:
        logger.info(f"Cache file {cache_file_path} does not exist.")
        return None

def save_cache(data: Any, cache_file_path: str) -> None:
    """
    Saves data to a cache file.

    Args:
        data (Any): Data to be cached.
        cache_file_path (str): Path to the cache file.
    """
    logger.info(f"Saving data to cache file {cache_file_path}.")
    try:
        with open(cache_file_path, 'wb') as cache_file:
            pickle.dump(data, cache_file)
        logger.info(f"Data cached successfully at {cache_file_path}.")
    except Exception as e:
        logger.error(f"Error saving cache to {cache_file_path}: {e}")

File: app/api/search/cache/cacheManager.ts
// File: app/api/search/cache/cacheManager.ts

import fs from 'fs';
import path from 'path';
import { promisify } from 'util';

const CACHE_DIRECTORY = './app/api/search/cache';
const readFile = promisify(fs.readFile);
const writeFile = promisify(fs.writeFile);

// Ensure the cache directory exists
export function ensureCacheDirectoryExists(): void {
  if (!fs.existsSync(CACHE_DIRECTORY)) {
    fs.mkdirSync(CACHE_DIRECTORY, { recursive: true });
    console.log(`Cache directory created at ${CACHE_DIRECTORY}.`);
  }
}

// Load data from cache
export async function loadCache<T>(cacheFilePath: string): Promise<T | null> {
  ensureCacheDirectoryExists();
  const fullPath = path.resolve(CACHE_DIRECTORY, cacheFilePath);

  if (fs.existsSync(fullPath)) {
    try {
      const data = await readFile(fullPath, 'utf-8');
      return JSON.parse(data) as T;
    } catch (error) {
      console.error(`Error loading cache from ${fullPath}: ${error}`);
      return null;
    }
  } else {
    return null;
  }
}

// Save data to cache
export async function saveCache<T>(data: T, cacheFilePath: string): Promise<void> {
  ensureCacheDirectoryExists();
  const fullPath = path.resolve(CACHE_DIRECTORY, cacheFilePath);
  try {
    await writeFile(fullPath, JSON.stringify(data, null, 2), 'utf-8');
    console.log(`Data cached successfully at ${fullPath}.`);
  } catch (error) {
    console.error(`Error saving cache to ${fullPath}: ${error}`);
  }
}

// Load cached data
export async function getCachedData(key: string): Promise<any> {
  const cacheFilePath = `${key}.json`; // Define how to derive the cache file name
  return await loadCache(cacheFilePath);
}

// Save cached data
export async function setCachedData(key: string, value: any): Promise<void> {
  const cacheFilePath = `${key}.json`; // Define how to derive the cache file name
  await saveCache(value, cacheFilePath);
}
